pybaseball/datahelpers/caching/func_utils.py

```diff
@@ -0,0 +1,71 @@
+import copy
+import hashlib
+import re
+from typing import Callable, Dict, Hashable, Iterable, Optional, Tuple, Union
+
+MAX_ARGS_KEY_LENGTH = 256
+
+def get_value_hash(value: Optional[Union[str, Dict, Hashable, Iterable]], include_designators: bool = True) -> str:
+    if value is None:
+        return 'None'
+
+    if isinstance(value, str):
+        # Remove invalid filename characters
+        stripped = copy.copy(value)
+        stripped = str(stripped).strip().replace(' ', '_')
+        stripped = re.sub(r'(?u)[^-\w.]', '', stripped)
+        if value == stripped:
+            # If nothing got changed, then the value is safe
+            if include_designators:
+                return f"'{value}'" if "'" not in value else f"\"{value}\""
+            return f"{value}"
+
+    if isinstance(value, dict):
+        values = ', '.join([
+            f"{get_value_hash(key, include_designators=False)}={get_value_hash(value[key])}"
+            for key in value.keys()
+        ]).strip(', ')
+        if include_designators:
+            return "{" + values + "}"
+        return values
+
+    if isinstance(value, tuple):
+        values = ', '.join([f"{get_value_hash(sub_value)}" for sub_value in value]).strip(', ')
+        if include_designators:
+            return f"({values})"
+        return f"{values}"
+
+    if isinstance(value, list):
+        values = ', '.join([f"{get_value_hash(sub_value)}" for sub_value in value]).strip(', ')
+        if include_designators:
+            return f"[{values}]"
+        return f"{values}"
+
+    try:
+        return str(value.__hash__())
+    except:
+        raise ValueError(f"value {value} of type {type(value)} is not hashable.")
+
+
+def get_func_name(func: Callable) -> str:
+    if '__self__' in dir(func):
+        # This is a class method
+        return f"{func.__getattribute__('__self__').__class__.__name__}.{func.__name__}"
+    return f"{func.__name__}"
+
+
+def get_func_hash(func: Callable, args: Tuple, kwargs: Dict) -> str:
+    f_hash = f"{get_func_name(func)}"
+
+    args_hash = get_value_hash(args, include_designators=False)
+    kwargs_hash = get_value_hash(kwargs, include_designators=False)
+
+    args_key = ', '.join([args_hash, kwargs_hash]).strip(', ')
+
+    # If the args_key is very long, just use a sha hash
+
+    if len(args_key) > MAX_ARGS_KEY_LENGTH:
+        # Let's hash this to make the filename shorter
+        args_key = hashlib.sha256(args_key.encode('utf-8')).hexdigest()
+
+    return f"{f_hash}({args_key})"
```

Comment:
why are we rolling our own hash?

Reply:
Because it generates something human readable if possible (such as `func_name(2019, keyword_param_name=True)`) so that the cache files can be viewed easily on the drive.

Reply:
Can we reduce this to along the lines
```python3
def get_func_hash(func: Callable, args: Tuple, kwargs: Dict) -> str:
    fname = get_func_name(func)
    args_s = str(args)
    kwargs_s = str(kwargs)
    args_key = f'{{args_s}--{kwargs_s}'
    # If the args_key is very long, just use a sha hash

    if len(args_key) > MAX_ARGS_KEY_LENGTH:
        # Let's hash this to make the filename shorter
        args_key = hashlib.sha1(args_key.encode('utf-8')).hexdigest()

    return f"{f_name}({args_key})"
```

Reply:
No for a few reasons:

1. If any of the args or kwargs has a character that does not belong in a filename (such as `/`)
2. `str(value)` is not guaranteed to be unique across objects. So a function call may inadvertently return a cached result despite never being called with the params it received.

Hence why I wrote the `get_value_hash` function to ensure safe parsing.

Reply:
Can you elaborate on the edge case in (2)? Because I would really like to try and figure out how we can avoid this system

Reply:
Sure. 2 is actually two separate issues. I'll give this code example to illustrate both:

```python
class A:
    def __init__(self, a, b):
        self.a = a
        self.b = b

x = A(1, 2)
print(str(x))

xy= A(1, 2)
print(str(y))
```

On my machine I get:
```
<__main__.A object at 0x000001C0435BEAC0>
<__main__.A object at 0x000001C0435BE340>
```

However, rerunning it, I get two entirely different results. So here's why that is important to us:

1. If another object generates that same value at some point (highly unlikely in this example, but not impossible, also is HIGHLY dependent on the implementation of `__str__` or `__repr__` for the value being passed to `str`).
2. More importantly, in this example we would want both `x` and `y` to be equivalent to our cache, but using `str` would mean we'd skip the cache as we'd see these as different values.

Now let's compare to what we generate (after I fixed a bug, and added a test to ensure this is true):
```
A({a=1, b=2})
A({a=1, b=2})
```

So this would generate a cache hit for each of those objects.

Reply:
if you noticed, I left in
```python3
    fname = get_func_name(func)
```
Since that object issue is important for functions.
However: do we use pure object code there? Do we even have anything along the lines of A() going into the hash?
As I see it, there are three components:
1. The function. `get_func_name` solves that string.
2. The args. Should be a list? Should be stringable in turn then, no?
3. The kwargs. Should be a dict, should be stringable too, no?

Reply:
@schorrm this is a fair point. Looking through the existing functions we'll be caching, all of them should be stringable (no classes).

At this point it's a matter of do we want the simple solution, or the future proof solution (in case we do want to cache a function that would take an object)? I could see an argument for either path.

I will say we would have to do something a little more complex than you suggested as we'll have to make sure there are no reserved filename characters in the generated file path.